HPCC-21259 generateGlobalId workunit option to generate GlobalIds

afishbeck opened the pull request:
If option generateGlobalId is set and a GlobalId is not sent via
HTTP header, then roxie will generate a new unique GlobalId.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-12311, label: autoGlobalId starts now.
The reason of this test is: Did not build yet.
Commit ID: 4d03051016f7d682114c3e0a01d383e5b2dc3eaf
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 4d03051016f7d682114c3e0a01d383e5b2dc3eaf
Number of Error(s): 0
CMake Error at cmake_modules/commonSetup.cmake:480 (add_library):
Number of Error(s): 0
CMakeFiles/CMakeOutput.log copied into ../CMakeOutput-2019-03-15_03-58-44.log
CMakeFiles/CMakeError.log copied into ../CMakeError-2019-03-15_03-58-44.log

[Build error:](https://gist.github.com/93c596be2cb6ce9603f9cda0ec2a012c)
Build log [Build log](https://gist.github.com/HPCCSmoketest/93c596be2cb6ce9603f9cda0ec2a012c#file-result-19-03-15-03-58-24-log)

richardkchapman commented:
@AttilaVamos Any ideas what failed on smoketest?

AttilaVamos commented:
@afishbeck The problem is:
```
CMake Error at cmake_modules/commonSetup.cmake:480 (add_library):
  Cannot find source file:

    /mnt/disk1/home/vamosax/smoketest/PR-12311/HPCC-Platform/system/globalid/lnuid.cpp
```
And if I see well, the lnuid.h also.
Somehow, the most important part is missing from the report. Please, give me some time to check and fix Smoketest.

richardkchapman commented:
I suspect Tony forgot to add them

@afishbeck

afishbeck commented:
Added missing files.  Waiting for smoke test to confirm.

HPCCSmoketest commented:
This PR (12311) is in the 2/2 position of the queue. Estimated start time is after ~1.0 hour(s)

HPCCSmoketest commented:
This PR (12311) is in the 1/1 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
Process of PR-12311, label: autoGlobalId starts now.
The reason of this test is: New commit.
Commit ID: 16364f38798369b4814efdd2d5a42e5323ade949
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 16364f38798369b4814efdd2d5a42e5323ade949
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |104 |104 |0 |0 |0 |33 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |819 |819 |0 |251 sec  (00:04:11) |
|test (thor) |744 |744 |0 |606 sec  (00:10:06) |
|test (roxie) |893 |893 |0 |347 sec  (00:05:47) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  195 sec (00:03:15) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1515 sec (00:25:15) |   19 sec (00:00:19) | 1774 sec (00:29:34) |

richardkchapman commented:
@wangkx Please review

wangkx commented on system/globalid/lnuid.h:
```diff
@@ -0,0 +1,59 @@
+    bool sameUniqueId(const ln_uid_t &uid1, const ln_uid_t &uid2);
+    void copyUniqueId(ln_uid_t &to, const ln_uid_t &from);
+
+	int get_utc_offset();
```
wrong indent?

afishbeck replied:
the indent issues may have been the way I copied Amol's code.  This code is used to generate IDs in LN we are open sourcing it.  I'll take a look at the original.

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        gettimeofday(&tv, NULL);
+
+        millisecondsSinceEpochTrim =
+        (unsigned long long int)(tv.tv_sec) * 1000 +
```
indent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+
+
+        for(unsigned int i=0, j=0; i <time_byte_count; i++, j++) {
+            out[i] = timedata[j];
```
Why do you need both i and j (they are the same)?

afishbeck replied:
Agreed.  Copied from Amol, but I'll change it.

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        // Skip leading zeroes in base58 result.
+        vector<unsigned char>::iterator it = b58.begin() + (size - length);
+        while (it != b58.end() && *it == 0)
+        it++;
```
indent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        str.reserve(zeroes + (b58.end() - it));
+        str.assign(zeroes, '1');
+        while (it != b58.end())
+        str += pszBase58[*(it++)];
```
indent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+
+        // Skip leading spaces. psz
+        while (*uid && isspace(*uid))
+        uid++;
```
intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+            // Decode base58 character
+            const char* ch = strchr(pszBase58, *uid);
+            if (ch == NULL)
+            return out;
```
intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        }
+        // Skip trailing spaces.
+        while (isspace(*uid))
+        uid++;
```
intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        while (isspace(*uid))
+        uid++;
+        if (*uid != 0)
+        return out;
```
intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        // Skip leading zeroes in b256.
+        vector<unsigned char>::iterator it = b256.begin();
+        while (it != b256.end() && *it == 0)
+        it++;
```
intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        vch.reserve(zeroes + (b256.end() - it));
+        vch.assign(zeroes, 0x00);
+        while (it != b256.end())
+        vch.push_back(*(it++));
```
intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+    void getUniqueIdRange(time_t start, time_t end, ln_uid_t &uid_start, ln_uid_t &uid_end){
+
+        unsigned long long  int startMilliSecondsSinceEpochTrim =
+        (unsigned long long int)(start) * 1000;
```
intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        unsigned long long  int startMilliSecondsSinceEpochTrim =
+        (unsigned long long int)(start) * 1000;
+        unsigned long long  int endMilliSecondsSinceEpochTrim =
+        (unsigned long long int)(end) * 1000;
```
Intent?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        }
+
+        for(unsigned int i=0, j=0; i <time_byte_count; i++, j++) {
+            uid_start[i] = startTimeData[j] ;
```
uid_start[i] = startTimeData[i];

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        }
+
+        for(unsigned int i=0, j=0; i <time_byte_count; i++, j++) {
+            uid_end[i] = endTimeData[j] ;
```
uid_end[i] = endTimeData[i] ;

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+#ifdef __linux__
+        localtime_r(&start_time, &tm);
+        sprintf(sb, "now: %4d-%02d-%02d %02d:%02d:%02d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+        tm.tm_hour, tm.tm_min, tm.tm_sec);
```
intent?

wangkx replied:
Missed this one?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+#elif _WIN32
+        localtime_s(&tm, &start_time);
+        sprintf_s(sb, "now: %4d-%02d-%02d %02d:%02d:%02d\n", tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday,
+        tm.tm_hour, tm.tm_min, tm.tm_sec);
```
intent?

wangkx replied:
Another one?

wangkx commented on system/globalid/lnuid.cpp:
```diff
@@ -0,0 +1,403 @@
+        /* if the local time is the "day before" the UTC, subtract 24 hours
+from the hours to get the UTC offset */
+        if (timeptr.tm_mday < 2)
+        gmtime_hours -= 24;
```
intent?

wangkx reviewed (commented):
@afishbeck several minor comments

afishbeck committed:
HPCC-21259 generateGlobalId workunit option to generate GlobalIds

If option generateGlobalId is set and a GlobalId is not sent via
HTTP header, then roxie will generate a new unique GlobalId.

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-12311, label: autoGlobalId starts now.
The reason of this test is: New commit.
Commit ID: b5ec6ef664d071287ec9454bcd6b03617f96f8e2
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b5ec6ef664d071287ec9454bcd6b03617f96f8e2
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/PR-12311/HPCC-Platform/roxie/ccd/ccdlistener.cpp:1401:48: error: no matching function for call to \xe2\x80\x98RoxieProtocolMsgContext::setTransactionId(const char*, bool)\xe2\x80\x99

[Build error:](https://gist.github.com/eb39514bc94984bfae9620645982e593)
Build log [Build log](https://gist.github.com/HPCCSmoketest/eb39514bc94984bfae9620645982e593#file-result-19-03-15-18-46-34-log)

afishbeck commented:
Need to fix that build issue due to rebasing onto prior change.

HPCCSmoketest commented:
Process of PR-12311, label: autoGlobalId starts now.
The reason of this test is: New commit.
Commit ID: d6b9d16f175917020d8426df74f8e0b12c833131
Estimated completion time is ~0.5 hour(s)

afishbeck commented:
Fixed rebase build issue.
@wangkx please re-review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: d6b9d16f175917020d8426df74f8e0b12c833131
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |110 |110 |0 |0 |0 |33 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |820 |820 |0 |244 sec  (00:04:04) |
|test (thor) |745 |745 |0 |598 sec  (00:09:58) |
|test (roxie) |894 |894 |0 |348 sec  (00:05:48) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   23 sec (00:00:23) |  216 sec (00:03:36) |    0 sec (00:00:00) |    2 sec (00:00:02) |   18 sec (00:00:18) | 1478 sec (00:24:38) |   20 sec (00:00:20) | 1757 sec (00:29:17) |

wangkx reviewed (commented):
@afishbeck 2 minors

HPCCSmoketest commented:
Process of PR-12311, label: autoGlobalId starts now.
The reason of this test is: New commit.
Commit ID: 43d417384cfb13d5474723a9432a702b6b3077ab
Estimated completion time is ~0.5 hour(s)

afishbeck commented:
@wangkx updated.  Please check again.

wangkx commented:
@afishbeck looks good.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 43d417384cfb13d5474723a9432a702b6b3077ab
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |110 |110 |0 |0 |0 |33 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |820 |820 |0 |254 sec  (00:04:14) |
|test (thor) |745 |745 |0 |597 sec  (00:09:57) |
|test (roxie) |894 |894 |0 |348 sec  (00:05:48) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   30 sec (00:00:30) |  217 sec (00:03:37) |    0 sec (00:00:00) |    2 sec (00:00:02) |   19 sec (00:00:19) | 1487 sec (00:24:47) |   19 sec (00:00:19) | 1774 sec (00:29:34) |

afishbeck commented:
@richardkchapman please merge.
